### DISCUSSION TITLE
Consider whether the compatibility layer should be available on standalone Product Collection blocks

### DISCUSSION THREAD

**Discussion by medavidallsop** [woocommerce/woocommerce#52915] [WooCommerce, Blocks, and Site Editing]
I have been reading up a lot on the Product Collection block, due to the upcoming soft deprecation of the older product blocks.

You can read some of my thoughts on this here: https://github.com/woocommerce/woocommerce/discussions/48814#discussioncomment-11228525

From that link you can see that there isn't a means of using some of the old PHP hooks with this (when the block is added to a page in a standalone context, not within a category template, search template, etc), as the compatibility layer doesn't work in this scenario.

@kmanijak on Slack, said: https://woocommercecommunity.slack.com/archives/C03MCL9RTAR/p1731410485932169?thread_ts=1731404943.379009&cid=C03MCL9RTAR

_"The Compatibility Layer only runs on the archive templates (Product Catalog, Products by Categories/Tags/Attributes or Product Search Results) but not on post or pages. These hooks were used in Classic Templates only available on those templates, hence the support in those specific places as well."_

Which I completely understand in regards to feature parity, however a standalone product collection block, IMO, is effectively an archive, albeit not in a template.

**I also think a really good point here, is that if you use the `[products]` shortcode in a standalone page, this is effectively the equivalent of using the product collections block on a standalone page, and those old hooks work in that shortcode - yet do not in the product collections block.**

Therefore I just wanted to open up a discussion on whether the compatibility layer should be included for this block in a standalone non-template context. This would mean if you want to add content/functionality based on the old hooks, you don't have to do anything, whereas if it remains as is now, you have to do some hacky stuff with `render_block_woocommerce/product-collection` and potentially duplicate code to run via that hook, to replicate what you are doing on the existing classic PHP hooks, and then also look at adding a condition within it to not do what it is doing twice when the block is on templated pages.

**Comment by leewillis77** [comment]
> Therefore I just wanted to open up a discussion on whether the compatibility layer should be included for this block in a standalone non-template context. 

I would wholeheartedly agree with this. The compatibility layer is not really that useful if it only works for _some_ of the situations the old approach worked for.

**Comment by sunyatasattva** [comment]
Thanks for exploring solutions @medavidallsop! I think there is a bit of a misunderstanding on what the Compatibility Layer is supposed to represent here, or at least my understanding of it.

We implemented the Compatibility Layer as a good faith effort to ease the transition to the block-based paradigms to existing stores, by ensuring that essential functionality won't break resulting in users being put off by blocks and trying to do our best to smoothen the transition.

What the Compatibility Layer **is not**, is a supported, encouraged and alternative way to improve functionality for developers. Perhaps we should make it somehow clear in the code, or with some deprecation warnings on those hooks. But we want to transition to a fully block-based approach. We don't want to encourage hybrid use of blocks and shortcodes, nor is in our plans to add any functionality to the old paradigm.

We'd love to encourage developers to, instead, start embracing the blocks paradigm. I can most certainly guarantee that your use-case would be smoother to implement—and, most importantly, support in the long-term—with a block-based approach. I see in the other thread that you are trying to add some text after each item: is there a reason why couldn't this be a block within the “Product Template” block?

If your use-case is currently not achievable within the block paradigm, I think it would be more interesting and forward-thinking for us to understand why not, or why is it more difficult than you'd like, and work on solutions that are anchored in our new approach, rather than supporting what is essentially a work-around.

We understand that the Product Collection and, in general, Woo Blocks are a relatively new concept for Woo developers, and we are really open to feedback on how we can improve them and provide a better experience, so that we can all try to work in the same direction (i.e. a “blockified” future), instead of pulling in different ones.

I'll also summon @joshuatf and perhaps @nerrad and @kmanijak here in case I misspoke on my understanding of the purpose of the Compatibility Layer in the context of the Product Collection.

**Comment by leewillis77** [comment]
> I can most certainly guarantee that your use-case would be smoother to implement—and, most importantly, support in the long-term—with a block-based approach. I see in the other thread that you are trying to add some text after each item: is there a reason why couldn't this be a block within the “Product Template” block?

If by 'other thread' you mean https://github.com/woocommerce/woocommerce/discussions/44776 then that wasn't @medavidallsop, it was me, and my problems with transitioning to blocks are covered in that thread, so any response you can offer there would be great, but we'll leave this thread for the compatibility layer discussion.

**Reply by sunyatasattva** [reply]
Sorry, I misread! I am joining that discussion as well.

**Comment by medavidallsop** [comment]
@sunyatasattva Thanks for your feedback.

> We implemented the Compatibility Layer as a good faith effort to ease the transition to the block-based paradigms to existing stores, by ensuring that essential functionality won't break resulting in users being put off by blocks and trying to do our best to smoothen the transition.

Yes and I am very happy it exists, but for this to truly ease the transition and not break functionality for some sites, the compatibility layer should be considered for inclusion on standalone product blocks outside block templates, without this it can stop existing functionality added via hooks from working in some scenarios. Which will have some impact on use of these blocks.

> I see in the other thread that you are trying to add some text after each item: is there a reason why couldn't this be a block within the “Product Template” block?

If this was a simple static piece of text yes you could, in the same way in a classic world you could add it into the PHP templates.

However, a plugin that previously used these hooks could be doing something far more complex than that, the content being output could be conditional based on the product data of each product in the loop, user capabilities, etc.

Let's say, you have developed a plugin that adds content after each product in a list, regardless of the templates/shortcode/block implementation used to display that list, and the content is conditional depending on each individual product's data.

Since WooCommerce product blocks were introduced, there are now multiple scenarios a developer needs to cater for to ensure that functionality works for every site (there is actually another if you consider supporting the soft deprecated old product blocks but we'll leave that out for this):

1. Classic themes using PHP templates and potentially [product] shortcodes in posts/pages/embedded in a template
2. Classic themes using PHP templates and potentially Product Collection blocks in posts/pages
3. Block themes using block based templates and potentially [product] shortcodes in posts/pages/embedded in a template
4. Block themes using block based templates and potentially Product Collection blocks in posts/pages

Classic themes and shortcodes aren't going away, and still make up the majority of WooCommerce websites I work on, so must remain supported.

Now, as the standard Woo product loop hooks have been around forever, and as product collection blocks are new, most existing plugins will already be using a hook like `woocommerce_after_shop_loop_item` to implement their functionality/content.

As the compatibility layer is available in a block template context this will mean that any existing functionality added via `woocommerce_after_shop_loop_item` will have this impact in the above scenarios:

1. Will work
2. Will work on the PHP templates, but NOT in Product Collections blocks in post/pages
3. Will work
4. Will work on the block based templates, but NOT in Product Collections blocks in post/pages

This means for websites with scenario 2 and 4 those are potentially impacted.

If the compatibility layer will not be available in a standalone block context, this means the plugin developer needs to add additional functionality to replicate the old hook method either via some PHP hackery or by adding some JS block implementation - whereas if it was available there wouldn't be a need to do anything immediately, and the functionality will not be broken for those effected sites.

Over time the developer of the plugin could then look at adding a more specific block based implementation, but these blocks are to be available and more widely used for all sites in around a months time.

It is also important to consider that there are scenarios where functionality a plugin provides doesn't warrant the ability to be manageable/moveable/open to removal/etc via a WYSIWYG editor.

**Comment by nerrad** [comment]
> I also think a really good point here, is that if you use the [products] shortcode in a standalone page, this is effectively the equivalent of using the product collections block on a standalone page...

While it may appear so on the surface, this isn't necessarily true. The `[products]` shortcode is _only_ visually customizable by code and the user of the shortcode is expected to understand what the potential attributes are, or write code to override the defaults (i.e. edit the relevant template). With the block and site editor, much of the visual customization is not possible via the editor itself and we want to ensure that the user experience for things extending a block retains the intuitiveness of both being able to see what is displayed on the front end in the editor along with providing the capability to customize those things. As we continue to work towards this future, there are _many_ (not all, we **do** watch for places we can keep existing hooks) cases where we simply cannot provide the best experience for users while retaining the existing paradigm of how the code modified that content. We don't intend to leave developers like yourself hanging though - we do want to provide docs, and "before/after" type content to help you accomplish similar things in the new paradigm. What helps us provide useful help here is to be specific about what hooks you used and what you used them for. For example, what does your plugin do? What is some example content it adds? You mentioned, 

> It is also important to consider that there are scenarios where functionality a plugin provides doesn't warrant the ability to be manageable/moveable/open to removal/etc via a WYSIWYG editor.

While I agree, it's hard to know what gap exists that we _must_ address without a concrete example. There may be already a solution (but not via the way it may currently be done). Do you have a specific example that you can share with us?

> However, a plugin that previously used these hooks could be doing something far more complex than that, the content being output could be conditional based on the product data of each product in the loop, user capabilities, etc.

I'm not sure if it would meet all your use cases, but for extremely bespoke content where you don't care about what the user sees in the editor, there are several filters exposed via the [`render_block`](https://github.com/WordPress/wordpress-develop/blob/bdeb21b2cb8471782bc3905cf9143757ae77db53/src/wp-includes/blocks.php#L2088) function that you can hook into to inject content. Do any of those meet your needs?

**Reply by medavidallsop** [reply]
Thanks for the reply @nerrad, I'll get back to you with replies on some of these points soon.

Im struggling to understand why the compatibility layer functionality is available if it only works in a template context, is this a technical limitation?

If it doesn't work on the block in both a template (cat, search) and standalone page (homepage, posts, etc) context, it just adds confusion for developers, and potentially users who will see content appearing in some pages but not others.

**Reply by nerrad** [reply]
> Im struggling to understand why the compatibility layer functionality is available if it only works in a template context, is this a technical limitation?

It's because the initial iteration of the block templates were essentially wrappers around the PHP output with no editor customization for the content. This enabled stores with WooCommerce active to switch to a block theme (or new stores starting with a block theme) to be able to modify the content _around_ the classic output and not have any existing functionality (including customizations) from those templates impacted. As we iterate and create blocks for the various template parts, we're not increasing the surface area of that compatibility layer because the risk for unintended breakage between a mix of classic extending vs users modifying their templates visually (both in Woo core and other customizations by plugins and extensions) increases significantly. Plus, in most cases, the user experience for extending WooCommerce functionality in the new paradigm is (or will be) more user-friendly than the current.

That's a _very_ short summary that doesn't cover things like how our work here informed the [template registration API](https://github.com/WordPress/gutenberg/pull/61577) we contributed to WordPress core (which shipped with WordPress 6.7!), or all the other nuances related to building for the new paradigm 😄.

**Reply by medavidallsop** [reply]
Thanks for the explanation, 

I'm just trying to figure out the best approach for plugins who want to ensure their functionality works across both classic and block themes.

Let's say you have something hooked on `woocommerce_after_shop_loop_item`

This remains for classic theme compatibility.

You then hook into the block via the new methods for block compatibility.

On block theme templated pages, as the compatibility layer is there, it's going to show the content from `woocommerce_after_shop_loop_item`, and from the new block stuff... unless i'm missing something?

How would you get around that scenario... as I don't believe you could use an `is_archive()` condition or similar on the classic hook function as that would be true on both classic archives and block archives...? Is there some kinda "if I'm not in X block" condition?

**Example:**

The following code will add a "content from hook" div via the`woocommerce_after_shop_loop_item` hook. This must exist to support classic themes.

The code will also add a "content from block" div via the `render_block_woocommerce/product-collection` filter, I've done it via the filter as its the easiest/fastest way to show this vs the other native block methods. This must exist to ensure "content from block" is displayed on product collection blocks outside site editor templates.

```
function hook_add_content_after_product() {

	echo '<div>content from hook</div>';

}
add_action( 'woocommerce_after_shop_loop_item', 'hook_add_content_after_product' );

function block_add_content_after_product( $block_content, $block ) {

    $is_product_collection_block = $block['attrs']['query']['isProductCollectionBlock'] ?? false;

    if ( $is_product_collection_block ) {

        $block_content = str_replace( '</li>', '<div>content from block</div></li>', $block_content ); // This is just for testing, obviously this code could cause issues as it is targeting any </li>

    }

    return $block_content;

}
add_filter( 'render_block_woocommerce/product-collection', 'block_add_content_after_product', 10, 2 );
```

After adding this code if you are on a block/FSE theme, if you look at a product collection on the homepage it will look like this, which is okay (content from block is displaying):

<img width="2230" alt="image" src="https://github.com/user-attachments/assets/d56d50b6-7893-4958-b8ba-ab1cb98d4189">

However if you look at the category pages, you get this, showing both the classic and block content:

<img width="1328" alt="image" src="https://github.com/user-attachments/assets/fc683958-2b47-4659-8719-a4f6417e147b">

Which is because the compatibility layer is running there.

So if the compatibility layer cannot be used on standalone product collection blocks, there needs to be a means of stopping functionality hooked via the classic hooks from running when in the latter scenario above for block/FSE themes, or you get the duplication.

**Reply by kmanijak** [reply]
Thank you for providing this scenario!

Getting back to initial proposal about enabling the Compatibility Layer across posts and pages, we're not planning to do so. Instead, we're looking towards block-oriented solution from Core. I'd like to quote @sunyatasattva from [another discussion](https://github.com/woocommerce/woocommerce/discussions/44776):

> Like I mentioned in https://github.com/woocommerce/woocommerce/discussions/52915#discussioncomment-11315935, we are fully committed to a block-based future. So, while I agree with your sentiment that a plugin activation requiring extra user steps is a degraded experience on their perspective, we should work on a block-based solution instead of holding on to legacy paradigms.

And that solution is Block Hooks API. Obviously, there's currently a limitation that Block Hooks API doesn't take effect in Post Content: https://core.trac.wordpress.org/ticket/61074.

I'm happy to see and share that there's some recent progress in Core in this area: https://github.com/WordPress/wordpress-develop/pull/7898. (The change could land in 6.8). And I really appreciate your input @medavidallsop, as your comments and discussion highlighted the importance of the issue and accelerated a work on it.

Block Hooks API also solves the problem if differentiating templates vs posts by providing a `$context` argument that:

> can be either a `WP_Block_Template`  or `WP_Post` object or an array if the anchor block is in a pattern. 

([from docs](https://developer.wordpress.org/news/2024/03/25/exploring-the-block-hooks-api-in-wordpress-6-5/)).

---

I'm planning to test and give feedback to the PR in Core to verify if it satisfies the needs. Feel free to do the same. 🙌

Also, in upcoming weeks I'm planning to provide some guidelines regarding extensibility with Product Collection in the four scenarios you mentioned elsewhere (appreciate that! It nicely presents the overall state):

> - Classic themes using PHP templates and potentially [product] shortcodes in posts/pages/embedded in a template
> - Classic themes using PHP templates and potentially Product Collection blocks in posts/pages
> - Block themes using block based templates and potentially [product] shortcodes in posts/pages/embedded in a template
> - Block themes using block based templates and potentially Product Collection blocks in posts/pages

**Reply by medavidallsop** [reply]
@kmanijak

> Also, in upcoming weeks I'm planning to provide some guidelines regarding extensibility with Product Collection in the four scenarios you mentioned elsewhere (appreciate that! It nicely presents the overall state):

Thank you that would be great to hear, and would ideally like to know this before the next Woo release which makes use of Product Collection blocks more prominent. In addition you may also want to include the details relating to the older Product Blocks as they are currently only being soft deprecated, so we need to continue support for those too.

In this [Slack chat](https://woocommercecommunity.slack.com/archives/C1KAZ91E3/p1732740834581769?thread_ts=1732723378.793919&cid=C1KAZ91E3) there was a potential workaround posted:

>If I'm understanding the question correctly, we encountered this already in one of our plugins and ended up using is_fse_theme() in the legacy hook to avoid running the rest of the code for the compatibility layer version - which I think is what you were looking for above. Then we had a way to provide a block based solution via other block filters and avoid the duplicate content. Anyone using the legacy hook through the normal non block based solutions would still carry on working the same way.

So I tried this and here are my findings:

>Hi, thats pretty useful, I see that I could use !wp_is_block_theme() to output the classic hook content not on block themes, however I have found some issues there, as it could be a block theme but they have used the [products] shortcode, and in that scenario the classic hooks wont work on it due to the condition. Now, you might be thinking thats a little edge case, however the scenario does exist. On my travels I have also found some plugins that output product lists can use the [products] shortcode as an easy means of displaying lists of products, usually conditionally amending the shortcode attributes to display certain products based on some custom conditions. So while that condition you mention does work, there is this scenario still existing which could potentially arise. Karol Manijak: with the Block Hooks API, once available for pages/posts, does this help resolve this issue though? It seems like even if using that, this duplication issue when attempting to support both classic hooks and block hooks would still exist? As the block templates will still include the classic hooks, so there must be some condition in the classic hooks to stop that, yet the !wp_is_block_theme() wouldn't be able to achieve it fully due to the scenario above.

Therefore if in your guidelines you are considering anything related to use of a `!wp_is_block_theme()` condition around the classic hook output, can you consider the above scenario in that too.

**Reply by leewillis77** [reply]
I've just had a play with this, and I'd like to echo @medavidallsop's concerns.

At the moment I can't see how to author an extension that adds in content and supports both:
- Sites with legacy archive layouts (using hooks like `woocommerce_after_shop_loop_item_title`)
- Sites with product collection based output (using block hooks)

It seems like we'll either get missing output or duplicate output depending on how we do things with no way to avoid it. I think we'll have to keep any work to make our extensions block-based on the backburner for now until this is all cleared up.

**Comment by kmanijak** [comment]
**TLDR: I gave it a lot of thoughts and I'm actually thinking we should enable Compatibility Layer across posts and pages as suggested in this discussion.**

---

I'd like to present @medavidallsop arguments but in more visual way to try convincing more folks.

Here are all the cases we're considering with the tools we currently have:

Keys:
- ✅ extensibility point works in that case.
- ❌ extensibility point doesn't work in that case.
- ⁉️ case in question.

<img width="1626" alt="image" src="https://github.com/user-attachments/assets/d52e7b33-d442-4e6e-ae20-78f44d17a52b">

**What's important to notice is:**
- The symmetry between left side (Archive Product) and right side (Posts/pages) for classic solutions - they take the same effect no matter the location.
- The asymmetry in case of Compatibility Layer.

## Use case 

Let's imagine plugin developer wants to include additional content after each product. Considering stores have different combinations plugins need to cover all the combinations.

### 1. Before Product Collection block

If store is all classic, you use `woocommerce_after_shop_loop_item` hook and that covers everything. If store uses legacy blocks the hook is not in use. However legacy blocks exposed `woocommerce_blocks_product_grid_item_html`. It requires more logic to find a right place but:
- hook takes effect in archive template AND in posts,
- both hooks are mutually exclusive. If you want to target classic product grid you use one, if block, you use another.

<img width="1240" alt="image" src="https://github.com/user-attachments/assets/61d63e9a-e900-4369-8bc0-44e6f8ef4df5">

### 2. With Product Collection - current state 

In the current state there's lack of symmetry between classic solutions and Product Collection. There's Compatibility Layer but it only takes effect in templates. Any combination that includes Product Collection usage on posts (even the fully block-based store) face the problem of asymmetric extensibility points.

So to include some content you could use dedicated hook like `render_block_woocommerce/product-collection` to find a right place and inject content (similar to legacy product grids and `woocommerce_blocks_product_grid_item_html`) . However there's a conflict:
- `woocommerce_after_shop_loop_item` hook takes effect in archive template only,
- `render_block_woocommerce/product-collection` takes effect in archive template AND in posts **resulting in duplicated content in archive templates.**

<img width="1236" alt="image" src="https://github.com/user-attachments/assets/10e47ba1-c3b3-41d0-ac87-6c57d01d6f17">

Solutions:
- ❌ Disabling Compatibility Layer - that would bring the symmetry to the hook usage but it's not the way to go. From plugin perspective disabling Compatibility Layer means it may break all the other plugins relying on it.
- ❓ Writing differentiating logic - one option would be to for developers to write an additional logic recognising where the Product Collection is being rendered and based on that guess whether Compatibility Layer is in use or not. It increases complexity quite a lot and does not feel fully reliable.
- ❌ Block Hooks API - they currently only takes effect in templates. So they can only be used as a duplication of Compatibility Layer.

### 3. Future: With Product Collection - assuming Block Hooks are already enabled in posts

There's ongoing work to enable Block Hooks on posts (https://github.com/WordPress/wordpress-develop/pull/7898). **Let's _assume_:**
- Block Hooks are enabled on posts already since WordPress 6.8 (it's just assumption!),
- plugin developers wrote the differentiating logic from previous scenario while waiting for Block Hooks.

Now, we want Block Hooks to be an ultimate solution - no doubts about that. In this scenario though, we're facing the same conflict as in previous scenario but twice:
1. Block Hooks can conflict with `woocommerce_after_shop_loop_item` hook based on Compatibility Layer so that requires additional differentiation logic to disable either of the solutions in templates
2. Block Hooks can conflict with `render_block_woocommerce/product-collection` and that requires decision probably based on Block Hooks API + the information if `render_block_woocommerce/product-collection` works or not depending on previous differentiation logic.

Long story short: it gets complex really fast.

<img width="1237" alt="image" src="https://github.com/user-attachments/assets/54d0dac0-0cdb-4719-a190-36292f77fe6f">

---

## Solution: Enable Compatibility Layer on posts

Let's see what happens if we enable Compatibility Layer on posts.

1. First of all, that satisfies plugins' extensibility needs in all combinations, provides the symmetry of the solutions and removes the need for "manual" implementations through `render_block_woocommerce/product-collection` hook.

<img width="1237" alt="image" src="https://github.com/user-attachments/assets/be90c431-e2f1-4768-b8d3-021b15fad82f">

2. Once Block Hooks API are enabled on posts providing unified outcomes across all stores the transition will be smooth as well. Plugin can rely either:
- on classic hooks in classic or mixed (classic + blocks) stores
- on Block Hooks on fully block stored (and since specific WordPress version)

The "decision" logic is minimalistic, simple and eventually allows for smooth transition from classic solution to block one because they cover the same cases.

<img width="1236" alt="image" src="https://github.com/user-attachments/assets/2b6acb10-4479-4c4d-8c6c-cea473ed9221">

---

## Summary:

**I think we should enable Compatibility Layer across posts and pages because:**
- If the Compatibility Layer purpose was to make the transition from classic to blocks smoother, making it work consistently across the store actually makes it smoother and more reliable.
- Seems like some plugins could be providing worse experience for the stores using Product Collection on posts and pages already and there's no easy and good alternative for plugin developers at the moment. Some solutions now increases the complexity now and for future transitions to Block Hooks API.
- It will make the future transition from Compatibility Layer to Block Hooks smoother.

---

## Tech notes:
1. I think it could be an opt-in solution so basically store owner could enable it on posts and pages.
2. Post Compatibility Layer would be separate from Archive Product and Single Product ones simply because not all hooks that work on archive templates work on posts and pages, for example:
    - `woocommerce_after_shop_loop_item` is applied to any product grid (in templates and posts),
    - `woocommerce_before_main_content` is applied only to the template.

---

cc: @gigitux, @imanish003, @sunyatasattva, @danielwrobert, @nerrad 

I hope this (probably too long 😅) comment presents the situation and actually explains my point how this move simplifies the whole situation and _actually_ is a future-proof move for blocks. I see the whole extensibility transition as two-steps process:

`1.` Classic hooks ⏯️ `2.` Compatibility Layer ➡️ `3.` Block Hooks API

With the eventual target solution in `3.`, we're currently somewhere between `1.` and `2.`. While this suggestion might seem like a step backward, I believe it’s actually **a corrective move,** helping us fully land on point 2.

I discussed the whole thing with @gigitux before writing this whole post and we both think an opt-in solution is a good way forward. However, it's a big thing so I'd love to hear your opinion on that.

_Disclaimer: I used Google Spreadsheet rather than markdown tables to have better control over colors and table layout._

**Reply by sunyatasattva** [reply]
Thanks @kmanijak for the thorough assessment of the situation. I understand your point and the user's concerns, and I also understand that—practically—the proposed solution seems to be the most direct one. You explain the problem of practical inconsistencies, while for me the issue is philosophical or conceptual inconsistencies on a product level.

Like [I mentioned in my other comments](https://github.com/woocommerce/woocommerce/discussions/52915#discussioncomment-11315935), I think what this discussion is really about is the nature and objective of the Compatibility Layer.

In my view, there continues to be a misunderstanding of its goal. When you say:

> If the Compatibility Layer purpose was to make the transition from classic to blocks smoother

I don't think the goal was to make it smoother in the sense of supporting and encouraging old practices and paradigms, and enabling developers to avoid the unease of getting out of the familiar comfort zone. We can make migration paths smooth and as easy as we can, and we should definitely focus a lot on DevEx, by figuring out the pain points. However, some work to migrate to the new paradigm will be required.

The full context of what I meant when I said “smoothen the transition” was:

>  a good faith effort to ease the transition to the block-based paradigms to existing stores, by **ensuring that essential functionality won't break** resulting in users being put off by blocks and trying to do our best to smoothen the transition.

Essentially, we wanted users to be able to opt in to blockified templates, without fear of breaking anything _existing_ and taking their store down.

The default experience of a blockified store is now to have the Product Collection in the archives. The Compatibility Layer ensures that the store keeps functioning as it was before.

Using the Product Collection on pages is _a new_ option and capability of our platform. Using it will require developers to adapt to the new paradigm: most of the use cases should be covered by adjusting one's design and architecture paradigm away from the old “hooks” and directly into “blocks”. I think that's the main struggle that most developers will face: the difficulty of adopting a block-first mentality. The rest, by block hooks, which, IMO, should be used sparingly.

This is not to invalidate your points: the use-cases exemplified here are completely valid, and the current limitations are a pain to deal with. However, my opinion stays in the camp of: we need to take this as an opportunity to improve whatever APIs are currently lacking within the context of the new paradigm. Otherwise we risk shooting ourselves in the foot by setting a precedent in using old-fashioned hooks as a solution _and_ diverging from the philosophy of Core.

I truly understand the pain and frustration that comes with this, but **I would like to see this conversation move more towards thinking of possible problems and solutions within the current direction and context**. Transitions are hard, but we're not doing anyone any favors by lingering in this limbo for longer.

“Hybrid” stores (mixing classic and block paradigms) shouldn't be encouraged as a solution. If anything, they are (should be) a temporary step towards a full migration to blocks, or we are preparing a recipe for a messy disaster, in my opinion.

**Reply by medavidallsop** [reply]
@kmanijak thanks so much for deciphering my previous messages and putting this into something more legible!

Based on all this, it feels like the only way to get around what has been discussed would be to either:

1. Enable the compatibility layer on posts/pages OR
2. Remove the compatibility layer entirely, so you can still use classic hooks for old, and whatever new method for product collection blocks, without any clashes/duplication/confusion

I must emphasise with 2, the inevitable "why don't my hooks work with these blocks" queries will come thick and fast, I've seen hundreds of these queries for any Woo functionality that gets blockified.

>Using the Product Collection on pages is a new option and capability of our platform.

@sunyatasattva It depends on how you gauge this, as in classic world you could use a `[products]` shortcode on a page happily and the usual loop based hooks would work for it (which effectively is the equivalent of a product collection block, I appreciate the management aspect differs, but the ultimate end result for both methods is you get a loop of products on a page)

**Reply by leewillis77** [reply]
@kmanijak Thanks for the detailed analysis, and for taking time to understand the issues.

I just wanted to flag that your solution proposal doesn't (as far as I can tell) address all of the problems.

Specifically, if an extension author tries to support both non-block stores, and block stores (which practically we have to), then there's no way to do so without generating duplicate output.

ie in your second solution table, if we implement a hook (woocommerce_after_shop_loop_item as an example) to support old stores and also provide blocks via the block hooks API to support blockified stores/templates/pages then we'll get duplicate output without any way of avoiding it?

As things stand, the best (in terms of providing the best quality experience for extension users) option for us in that case is to "do nothing" and rely on the legacy hooks and avoid moving to blocks. That seems probably not the behaviour that you'd like to incentivise.

    while for me the issue is philosophical or conceptual inconsistencies on a product level.

@sunyatasattva This is definitely not philosophical, or conceptual. This is a real, practical problem affecting extension developers...

**Reply by sunyatasattva** [reply]
> @sunyatasattva This is definitely not philosophical, or conceptual. This is a real, practical problem affecting extension developers...

@leewillis77 there was a misunderstanding in my words. I have, in fact, said:

> the use-cases exemplified here are completely valid, and the current limitations are a pain to deal with.

I recognize the situation as problematic and very real. This is why I am engaging in this conversation. What I mean when I say that the issue here is conceptual is that the solution is not just “let's extend the compatibility layer and call it a day”: this solution would not be aligned with the product direction, and—while perhaps solving a specific pain point you are experiencing—will end up creating more problems down the line (whether it means tech debt, inconsistencies, precedents, conflicts with our own goals of increase block theme adoption, etc.).

I insist in saying that what I would like to see for this conversation to be productive, is thinking about a solution that **both** solves your problem and use-case, **and** aligns with the direction we are going. I think such solution is possible and that's where we should focus our efforts.
